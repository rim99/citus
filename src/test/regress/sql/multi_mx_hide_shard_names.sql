--
-- Hide shard names on MX worker nodes
--

SET citus.next_shard_id TO 1130000;


CREATE SCHEMA mx_hide_shard_names;
SET search_path TO 'mx_hide_shard_names';

SET citus.shard_count TO 4;
SET citus.shard_replication_factor TO 1;

SET citus.replication_model TO 'streaming';
SELECT start_metadata_sync_to_node('localhost', :worker_1_port);
SELECT start_metadata_sync_to_node('localhost', :worker_2_port);

CREATE TABLE test_table(id int, time date);
SELECT create_distributed_table('test_table', 'id');

-- first show that the views does not show
-- any shards on the coordinator as expected
SELECT * FROM citus_shards_on_worker;
SELECT * FROM citus_shard_indexes_on_worker;

-- now show that we see the shards, but not the 
-- indexes as there are no indexes
\c - - - :worker_1_port
SET search_path TO 'mx_hide_shard_names';
SELECT * FROM citus_shards_on_worker ORDER BY 2;
SELECT * FROM citus_shard_indexes_on_worker ORDER BY 2;

-- now create an index
\c - - - :master_port
SET search_path TO 'mx_hide_shard_names';
CREATE INDEX test_index ON mx_hide_shard_names.test_table(id);

-- now show that we see the shards, and the 
-- indexes as well
\c - - - :worker_1_port
SET search_path TO 'mx_hide_shard_names';
SELECT * FROM citus_shards_on_worker ORDER BY 2;
SELECT * FROM citus_shard_indexes_on_worker ORDER BY 2;

-- we should be able to select from the shards directly if we 
-- know the name of the tables
SELECT count(*) FROM test_table_1130000;

-- disable the config so that table becomes visible
SELECT pg_table_is_visible('test_table_1130000'::regclass);
SET citus.enable_replace_table_visible_function TO FALSE;
SELECT pg_table_is_visible('test_table_1130000'::regclass);

\c - - - :master_port
-- make sure that we're resilient to the edge cases
-- such that the table name includes the shard number
SET search_path TO 'mx_hide_shard_names';
SET citus.shard_count TO 4;
SET citus.shard_replication_factor TO 1;

SET citus.replication_model TO 'streaming';
CREATE TABLE test_table_102008(id int, time date);
SELECT create_distributed_table('test_table_102008', 'id');

\c - - - :worker_1_port
SET search_path TO 'mx_hide_shard_names';
SELECT * FROM citus_shards_on_worker ORDER BY 2;


\c - - - :master_port
-- make sure that don't mess up with schemas
CREATE SCHEMA mx_hide_shard_names_2;
SET search_path TO 'mx_hide_shard_names_2';
SET citus.shard_count TO 4;
SET citus.shard_replication_factor TO 1;

SET citus.replication_model TO 'streaming';
CREATE TABLE test_table(id int, time date);
SELECT create_distributed_table('test_table', 'id');
CREATE INDEX test_index ON mx_hide_shard_names_2.test_table(id);

\c - - - :worker_1_port
SET search_path TO 'mx_hide_shard_names';
SELECT * FROM citus_shards_on_worker ORDER BY 2;
SELECT * FROM citus_shard_indexes_on_worker ORDER BY 2;
SET search_path TO 'mx_hide_shard_names_2';
SELECT * FROM citus_shards_on_worker ORDER BY 2;
SELECT * FROM citus_shard_indexes_on_worker ORDER BY 2;
SET search_path TO 'mx_hide_shard_names_2, mx_hide_shard_names';
SELECT * FROM citus_shards_on_worker ORDER BY 2;
SELECT * FROM citus_shard_indexes_on_worker ORDER BY 2;

-- clean-up
\c - - - :master_port


-- show that common psql functions do not show shards
-- including the ones that are not in the current schema
SET search_path TO 'mx_hide_shard_names';
\d
\di

DROP SCHEMA mx_hide_shard_names CASCADE;
DROP SCHEMA mx_hide_shard_names_2 CASCADE;
