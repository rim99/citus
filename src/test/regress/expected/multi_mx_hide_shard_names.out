--
-- Hide shard names on MX worker nodes
--
SET citus.next_shard_id TO 1130000;
CREATE SCHEMA mx_hide_shard_names;
SET search_path TO 'mx_hide_shard_names';
SET citus.shard_count TO 4;
SET citus.shard_replication_factor TO 1;
SET citus.replication_model TO 'streaming';
SELECT start_metadata_sync_to_node('localhost', :worker_1_port);
 start_metadata_sync_to_node 
-----------------------------
 
(1 row)

SELECT start_metadata_sync_to_node('localhost', :worker_2_port);
 start_metadata_sync_to_node 
-----------------------------
 
(1 row)

CREATE TABLE test_table(id int, time date);
SELECT create_distributed_table('test_table', 'id');
 create_distributed_table 
--------------------------
 
(1 row)

-- first show that the views does not show
-- any shards on the coordinator as expected
SELECT * FROM citus_shards_on_worker;
 Schema | Name | Type | Owner 
--------+------+------+-------
(0 rows)

SELECT * FROM citus_shard_indexes_on_worker;
 Schema | Name | Type | Owner | Table 
--------+------+------+-------+-------
(0 rows)

-- now show that we see the shards, but not the 
-- indexes as there are no indexes
\c - - - :worker_1_port
SET search_path TO 'mx_hide_shard_names';
SELECT * FROM citus_shards_on_worker ORDER BY 2;
       Schema        |        Name        | Type  |  Owner   
---------------------+--------------------+-------+----------
 mx_hide_shard_names | test_table_1130000 | table | postgres
 mx_hide_shard_names | test_table_1130002 | table | postgres
(2 rows)

SELECT * FROM citus_shard_indexes_on_worker ORDER BY 2;
 Schema | Name | Type | Owner | Table 
--------+------+------+-------+-------
(0 rows)

-- now create an index
\c - - - :master_port
SET search_path TO 'mx_hide_shard_names';
CREATE INDEX test_index ON mx_hide_shard_names.test_table(id);
-- now show that we see the shards, and the 
-- indexes as well
\c - - - :worker_1_port
SET search_path TO 'mx_hide_shard_names';
SELECT * FROM citus_shards_on_worker ORDER BY 2;
       Schema        |        Name        | Type  |  Owner   
---------------------+--------------------+-------+----------
 mx_hide_shard_names | test_table_1130000 | table | postgres
 mx_hide_shard_names | test_table_1130002 | table | postgres
(2 rows)

SELECT * FROM citus_shard_indexes_on_worker ORDER BY 2;
       Schema        |        Name        | Type  |  Owner   |       Table        
---------------------+--------------------+-------+----------+--------------------
 mx_hide_shard_names | test_index_1130000 | index | postgres | test_table_1130000
 mx_hide_shard_names | test_index_1130002 | index | postgres | test_table_1130002
(2 rows)

-- we should be able to select from the shards directly if we 
-- know the name of the tables
SELECT count(*) FROM test_table_1130000;
 count 
-------
     0
(1 row)

-- disable the config so that table becomes visible
SELECT pg_table_is_visible('test_table_1130000'::regclass);
 pg_table_is_visible 
---------------------
 f
(1 row)

SET citus.enable_replace_table_visible_function TO FALSE;
SELECT pg_table_is_visible('test_table_1130000'::regclass);
 pg_table_is_visible 
---------------------
 t
(1 row)

\c - - - :master_port
-- make sure that we're resilient to the edge cases
-- such that the table name includes the shard number
SET search_path TO 'mx_hide_shard_names';
SET citus.shard_count TO 4;
SET citus.shard_replication_factor TO 1;
SET citus.replication_model TO 'streaming';
CREATE TABLE test_table_102008(id int, time date);
SELECT create_distributed_table('test_table_102008', 'id');
 create_distributed_table 
--------------------------
 
(1 row)

\c - - - :worker_1_port
SET search_path TO 'mx_hide_shard_names';
SELECT * FROM citus_shards_on_worker ORDER BY 2;
       Schema        |           Name           | Type  |  Owner   
---------------------+--------------------------+-------+----------
 mx_hide_shard_names | test_table_102008_102012 | table | postgres
 mx_hide_shard_names | test_table_102008_102014 | table | postgres
 mx_hide_shard_names | test_table_1130000       | table | postgres
 mx_hide_shard_names | test_table_1130002       | table | postgres
(4 rows)

\c - - - :master_port
-- make sure that don't mess up with schemas
CREATE SCHEMA mx_hide_shard_names_2;
SET search_path TO 'mx_hide_shard_names_2';
SET citus.shard_count TO 4;
SET citus.shard_replication_factor TO 1;
SET citus.replication_model TO 'streaming';
CREATE TABLE test_table(id int, time date);
SELECT create_distributed_table('test_table', 'id');
 create_distributed_table 
--------------------------
 
(1 row)

CREATE INDEX test_index ON mx_hide_shard_names_2.test_table(id);
\c - - - :worker_1_port
SET search_path TO 'mx_hide_shard_names';
SELECT * FROM citus_shards_on_worker ORDER BY 2;
       Schema        |           Name           | Type  |  Owner   
---------------------+--------------------------+-------+----------
 mx_hide_shard_names | test_table_102008_102012 | table | postgres
 mx_hide_shard_names | test_table_102008_102014 | table | postgres
 mx_hide_shard_names | test_table_1130000       | table | postgres
 mx_hide_shard_names | test_table_1130002       | table | postgres
(4 rows)

SELECT * FROM citus_shard_indexes_on_worker ORDER BY 2;
       Schema        |        Name        | Type  |  Owner   |       Table        
---------------------+--------------------+-------+----------+--------------------
 mx_hide_shard_names | test_index_1130000 | index | postgres | test_table_1130000
 mx_hide_shard_names | test_index_1130002 | index | postgres | test_table_1130002
(2 rows)

SET search_path TO 'mx_hide_shard_names_2';
SELECT * FROM citus_shards_on_worker ORDER BY 2;
        Schema         |       Name        | Type  |  Owner   
-----------------------+-------------------+-------+----------
 mx_hide_shard_names_2 | test_table_102016 | table | postgres
 mx_hide_shard_names_2 | test_table_102018 | table | postgres
(2 rows)

SELECT * FROM citus_shard_indexes_on_worker ORDER BY 2;
        Schema         |       Name        | Type  |  Owner   |       Table       
-----------------------+-------------------+-------+----------+-------------------
 mx_hide_shard_names_2 | test_index_102016 | index | postgres | test_table_102016
 mx_hide_shard_names_2 | test_index_102018 | index | postgres | test_table_102018
(2 rows)

SET search_path TO 'mx_hide_shard_names_2, mx_hide_shard_names';
SELECT * FROM citus_shards_on_worker ORDER BY 2;
 Schema | Name | Type | Owner 
--------+------+------+-------
(0 rows)

SELECT * FROM citus_shard_indexes_on_worker ORDER BY 2;
 Schema | Name | Type | Owner | Table 
--------+------+------+-------+-------
(0 rows)

-- clean-up
\c - - - :master_port
-- show that common psql functions do not show shards
-- including the ones that are not in the current schema
SET search_path TO 'mx_hide_shard_names';
\d
                     List of relations
       Schema        |       Name        | Type  |  Owner   
---------------------+-------------------+-------+----------
 mx_hide_shard_names | test_table        | table | postgres
 mx_hide_shard_names | test_table_102008 | table | postgres
(2 rows)

\di
                        List of relations
       Schema        |    Name    | Type  |  Owner   |   Table    
---------------------+------------+-------+----------+------------
 mx_hide_shard_names | test_index | index | postgres | test_table
(1 row)

DROP SCHEMA mx_hide_shard_names CASCADE;
NOTICE:  drop cascades to 2 other objects
DETAIL:  drop cascades to table test_table
drop cascades to table test_table_102008
DROP SCHEMA mx_hide_shard_names_2 CASCADE;
NOTICE:  drop cascades to table mx_hide_shard_names_2.test_table
